skills/contextual-embeddings/guide.ipynb

```diff
@@ -0,0 +1,1372 @@
+      "Evaluating retrieval: 100%|██████████| 248/248 [00:06<00:00, 39.43it/s]"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Pass@20: 90.06%\n",
+      "Total Score: 0.9006336405529954\n",
+      "Total queries: 248\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "results5 = evaluate_db(base_db, 'data/evaluation_set.jsonl', 5)\n",
+    "results10 = evaluate_db(base_db, 'data/evaluation_set.jsonl', 10)\n",
+    "results20 = evaluate_db(base_db, 'data/evaluation_set.jsonl', 20)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Contextual Embeddings\n",
+    "\n",
+    "With basic RAG, each embedded chunk contains a potentially useful piece of information, but these chunks lack context. With Contextual Embeddings, we create a variation on the embedding itself by adding more context to each text chunk before embedding it. Specifically, we use Claude to create a concise context that explains the chunk using the context of the overall document. In the case of our codebases dataset, we can provide both the chunk and the full file that each chunk was found within to an LLM, then produce the context. Then, we will combine this 'context' and the raw text chunk together into a single text block prior to creating each embedding.\n",
+    "\n",
+    "### Additional Considerations: Cost and Latency\n",
+    "\n",
+    "The extra work we're doing to 'situate' each document happens only at ingestion time: it's a cost you'll pay once when you store each document (and periodically in the future if you have a knowledge base that updates over time). There are many approaches like HyDE (hypothetical document embeddings) which involve performing steps to improve the representation of the query prior to executing a search. These techniques have shown to be moderately effective, but they add significant latency at runtime.\n",
+    "\n",
+    "Prompt caching also makes this much more cost effective. Creating contextual embeddings requires us to pass the same document to the model for every chunk we want to generate extra context for. With prompt caching, we can write the overall doc to the cache once, and then because we're doing our ingestion job all in sequence, we can just read the document from cache as we generate context for each chunk within that document (the information you write to the cache has a 5 minute time to live). This means that the first time we pass a document to the model, we pay a bit more to write it to the cache, but for each subsequent API call that contains that doc, we receive  a 90% discount on all of the input tokens read from the cache. Assuming 800 token chunks, 8k token documents, 50 token context instructions, and 100 tokens of context per chunk, the cost to generate contextualized chunks is $1.02 per million document tokens.\n",
```

Comment:
It might be worth clarifying here how the cached document -> chunk relationship works. I.e. if you cache the whole document but then run inference to generate the contextual summarization on just the chunk, how does that relate to what has been cached?